libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LocationValidator.java

```diff
@@ -0,0 +1,123 @@
+package com.mapbox.services.android.navigation.v5.location;
+
+import android.location.Location;
+import android.support.annotation.NonNull;
+
+import com.mapbox.geojson.Point;
+import com.mapbox.turf.TurfConstants;
+import com.mapbox.turf.TurfMeasurement;
+
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.TimeUnit;
+
+public class LocationValidator {
+
+  private static final int TIME_THRESHOLD = 5000; // 5 seconds
+  private static final int ACCURACY_THRESHOLD = 10; // percent
+  private static final int VELOCITY_THRESHOLD = 200; // milliseconds
+
+  private static LocationValidator instance;
+  private Location lastValidLocation;
+
+  private final CopyOnWriteArrayList<OnLocationLoggedListener> onLocationLoggedListeners
+    = new CopyOnWriteArrayList<>();
+
+  /**
+   * Primary access method (using singleton pattern)
+   *
+   * @return LocationValidator
+   */
+  public static synchronized LocationValidator getInstance() {
+    if (instance == null) {
+      instance = new LocationValidator();
+    }
+
+    return instance;
+  }
+
+  private LocationValidator() {
+  }
+
+  public boolean isValidUpdate(@NonNull Location location) {
+    // First update
+    if (lastValidLocation == null) {
+      lastValidLocation = location;
+      return true;
+    }
+
+    LoggingLocation loggingLocation = new LoggingLocation();
+
+    float currentAccuracy = location.getAccuracy();
+    float previousAccuracy = lastValidLocation.getAccuracy();
+    float accuracyDifference = Math.abs(previousAccuracy - currentAccuracy);
+
+    loggingLocation.setAccuracy(String.format("Accuracy - Current: %s, Previous: %s, Difference: %s",
```

Comment:
BTW, why setting a formatted `String` instead of setting the primitive values into `LoggingLocation`? We can always convert it to `String` afterwards. The same comment applies to `worseAccuracyAcceptable`, `distanceBetweenUpdates`, `timeSinceLastUpdate`, `velocity`, `validVelocity` and `validLocation`.